include/deal.II/base/memory_space.h

```diff
@@ -0,0 +1,212 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_memory_space_h
+#define dealii_memory_space_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/cuda.h>
+#include <deal.II/base/exceptions.h>
+
+#include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ *
+ */
+namespace MemorySpace
+{
+  /**
+   * Structure describing Host memory space.
+   */
+  struct Host
+  {};
+
+
+
+  /**
+   * Structure describing CUDA memory space.
+   */
+  struct CUDA
+  {};
+
+
+
+#if defined(DEAL_II_WITH_CUDA) && defined(__CUDACC__)
+  /**
+   * Deleter for unique/shared pointer pointing to an array allocated on the
+   * device.
+   */
+  template <typename Number>
+  void
+  deleter_device_array(Number *ptr) noexcept
+  {
+    const cudaError_t cuda_error_code = cudaFree(ptr);
+    (void)cuda_error_code;
+    AssertNothrow(cuda_error_code == cudaSuccess,
+                  ExcCudaError(cudaGetErrorString(cuda_error_code)));
+  }
+#endif
+
+
+
+  /**
+   * Data structure
+   */
+  template <typename Number, typename MemorySpace>
+  struct MemorySpaceData
+  {
+    MemorySpaceData()
+    {
+      static_assert(std::is_same<MemorySpace, Host>::value ||
+                      std::is_same<MemorySpace, CUDA>::value,
+                    "MemorySpace should be Host or CUDA");
+    }
+
+    /**
+     * Copy the active data (values for Host and values_dev for CUDA) to @p begin.
+     * If the data is on the device it is moved to the host.
+     */
+    void
+    copy_to(Number *begin, std::size_t n_elements)
+    {}
+
+    void
+    /**
+     * Copy the data in @p begin to the active data of the structure (values for
+     * Host and values_dev for CUDA). The pointer @p begin must be on the host.
+     */
+    copy_from(Number *begin, std::size_t n_elements)
+    {}
+
+    /**
+     * Pointer to data on the host.
+     */
+    std::unique_ptr<Number[], decltype(&free)> values;
+
+    /**
+     * Pointer to data on the device.
+     */
+    std::unique_ptr<Number[]> values_dev;
+  };
+
+
+
+  /**
+   * Swap function similar to std::swap.
+   */
+  template <typename Number, typename MemorySpace>
+  inline void
+  swap(MemorySpaceData<Number, MemorySpace> &,
+       MemorySpaceData<Number, MemorySpace> &)
+  {}
```

Comment:
I suggest to either only declare the template or put a `AssertThrow(false,...)` in.